jujugui/static/gui/src/app/store/env/base.js

```diff
@@ -294,30 +298,39 @@ YUI.add('juju-env-base', function(Y) {
       Close the WebSocket connection to the Juju API server.
 
       @method close
-      @param cb Optional callback.
+      @param callback Optional callback called once the connection has been
+        closed and the API cleaned up.
     */
-    close: function(cb) {
-      if (this.ws) {
-        this.beforeClose(this.ws.close.bind(this.ws));
+    close: function(callback) {
+      console.log(`closing the ${this.name} API connection`);
+      if (!callback) {
+        callback = () => {};
       }
-      if (cb) {
-        cb();
+      if (!this.ws) {
+        callback();
+        return;
       }
+      this.cleanup(() => {
+        this.userIsAuthenticated = false;
+        this.setCredentials(null);
+        this.ws.close();
+        callback();
+      });
     },
 
     /**
-      Define optional operations to be performed before closing the WebSocket
-      connection. This method, as defined here, only calls the given callback
-      as it is intended to be overridden by subclasses. Implementations are
-      responsible of calling the given callback that effectively closes the
-      WebSocket connection.
-
-      @method beforeClose
-      @param {Function} callback A callable that must be called by the
-        function and that actually closes the connection.
+      Define optional operations to be performed before logging out. This
+      method, as defined here, only calls the given callback as it is intended
+      to be overridden by subclasses. Implementations are responsible of
+      calling the given callback that effectively closes the WebSocket
+      connection. Concrete implementations are assumed to be idempotent.
+
+      @method cleanup
+      @param {Function} done A callable that must be called by the function and
+        that actually closes the connection.
     */
-    beforeClose: function(callback) {
-      callback();
+    cleanup: function(done) {
```

Comment:
Well I renamed those to make it more apparent that this is not something you could (or could not) call, like a usual callback. This is something that you are required to call in order to communicate that you are "done" with the clean up process. How does it sound?